File: n_queens.py
import copy

from typing import List

def n_queens(n: int) -> int:
    def make_board(n: int) -> List[List[str]]:
            return [['.'] * n for i in range(n)]
        
    def diag(row: int, col: int) -> int:
        return n - 1 + row - col
    
    def anti(row: int, col: int) -> int:
        return row + col
    
    def put_queen(row: int, col: int) -> None:
        cols[col] = False
        diags[diag(row, col)] = False
        antis[anti(row, col)] = False
        board[row][col] = 'Q'
        
    def remove_queen(row: int, col: int) -> None:
        cols[col] = True
        diags[diag(row, col)] = True
        antis[anti(row, col)] = True
        board[row][col] = '.'
        
    def can_put_queen(row: int, col: int) -> bool:
        return cols[col] and diags[diag(row, col)] and antis[anti(row, col)]
    
    def worker(no_queens: int, row: int) -> None:
        if no_queens:
            for col in range(n):
                if can_put_queen(row, col):
                    put_queen(row, col)
                    worker(no_queens - 1, row + 1)
                    remove_queen(row, col)
        else:
            for row in board:
                print(' '.join(row))
            print()
            # solutions.append([''.join(board_row) for board_row in board])

    cols = [True] * n
    diags = [True] * (2 * n - 1)
    antis = [True] * (2 * n - 1)
    board = make_board(n)
    solutions = []
    worker(n, 0)
    
    for board in solutions:
        for row in board:
            print(' '.join(row))
        print()
    
    return solutions


n_queens(15)

File: n_queens.cpp
#include <functional>
#include <iostream>
#include <vector>

void n_queens(int n)
{
    std::vector<bool> cols(n, true);
    std::vector<bool> diags(2 * n - 1, true);
    std::vector<bool> antis(2 * n - 1, true);
    std::vector<std::vector<char>> board(n, std::vector<char>(n, '.'));

    std::function<int(int, int)> diag = [&](int row, int col)
    {
        return n + row - col - 1;
    };

    std::function<int(int, int)> anti = [&](int row, int col)
    {
        return row + col;
    };

    std::function<bool(int, int)> can_put = [&](int row, int col)
    {
        return cols[col] && diags[diag(row, col)] && antis[anti(row, col)];
    };

    std::function<void(int, int)> put_queen = [&](int row, int col)
    {
        cols[col] = false;
        diags[diag(row, col)] = false;
        antis[anti(row, col)] = false;
        board[row][col] = 'Q';
    };

    std::function<void(int, int)> remove_queen = [&](int row, int col)
    {
        cols[col] = true;
        diags[diag(row, col)] = true;
        antis[anti(row, col)] = true;
        board[row][col] = '.';
    };

    std::function<std::string()> format_board = [&]()
    {
        std::string str(2 * n * n, '?');
        int i = -1;
        for (int row = n - 1; row > -1; --row)
        {
            for (int col = 0; col < n; ++col)
            {
                str[++i] = board[row][col];
                str[++i] = (col < n - 1) ? ' ' : '\n';
            }
        }
        return str;
    };

    std::function<void(int, int)> worker = [&](int no_queens, int row)
    {
        if (no_queens)
        {
            for (int col = 0; col < n; ++col)
            {
                if (can_put(row, col))
                {
                    put_queen(row, col);
                    worker(no_queens - 1, row + 1);
                    remove_queen(row, col);
                }
            }
        }
        else
        {
            std::cout << format_board() << '\n';
        }
    };
    worker(n, 0);
}

int main(int argc, char* argv[])
{
    for (size_t i = 1; i < argc; ++i)
    {
        n_queens(std::stoi(argv[i]));
    }
}